File: agent/src/agent/embed_client.py
"""Embed a single query → vector for Qdrant search. Mirrors indexer EMBED_MOCK behavior."""
from __future__ import annotations

import hashlib

from openai import AsyncOpenAI

from agent.config import settings
from agent.logging_setup import log

_DIM = 1536  # text-embedding-3-small — must match indexer/Qdrant collection.


class QueryEmbedClient:
    def __init__(self) -> None:
        self._mock = settings.embed_mock or not settings.openai_api_key
        if self._mock:
            log.info("embed.mock_mode", reason="EMBED_MOCK or no API key")
            self._client: AsyncOpenAI | None = None
        else:
            self._client = AsyncOpenAI(
                api_key=settings.openai_api_key,
                base_url=settings.openai_base_url,
            )

    async def embed(self, text: str) -> list[float]:
        if self._mock:
            return _mock_vector(text)
        assert self._client is not None
        resp = await self._client.embeddings.create(
            model=settings.embed_model, input=[text]
        )
        return resp.data[0].embedding


def _mock_vector(text: str) -> list[float]:
    h = hashlib.sha256(text.encode("utf-8")).digest()
    raw = (h * ((_DIM // len(h)) + 1))[:_DIM]
    return [(b - 128) / 128.0 for b in raw]

File: web/src/lib/kb/embed-client.ts
/**
 * Embed client — OpenAI-compat embeddings via native fetch.
 *
 * Env vars:
 *   OPENAI_API_KEY     — required for real mode
 *   OPENAI_BASE_URL    — default https://api.openai.com/v1
 *   EMBED_MODEL        — default text-embedding-3-small (dim 1536)
 *   KB_LLM_MOCK        — if "true" or OPENAI_API_KEY missing → sha256 mock vector
 *
 * Mock vector mirrors agent/src/agent/embed_client.py: sha256(text) → float[]
 * so dev/CI never needs a real API key.
 */

import { createHash } from "crypto";

const DIM = 1536; // text-embedding-3-small — must match Qdrant collection

function isMockMode(): boolean {
  return process.env.KB_LLM_MOCK === "true" || !process.env.OPENAI_API_KEY;
}

/** Deterministic mock vector from sha256(text), normalized to [-1, 1]. Mirrors Python impl. */
function mockVector(text: string): number[] {
  const hash = createHash("sha256").update(text, "utf8").digest();
  // Repeat hash bytes until we have DIM bytes, then map [0,255] → [-1,1]
  const repeated = Buffer.alloc(DIM);
  for (let i = 0; i < DIM; i++) {
    repeated[i] = hash[i % hash.length];
  }
  return Array.from(repeated).map((b) => (b - 128) / 128.0);
}

/**
 * Embed a single text string → float[1536].
 * Throws on API error (non-2xx response).
 */
export async function embedText(text: string): Promise<number[]> {
  if (isMockMode()) {
    return mockVector(text);
  }

  const baseUrl = (process.env.OPENAI_BASE_URL ?? "https://api.openai.com/v1").replace(/\/$/, "");
  const model = process.env.EMBED_MODEL ?? "text-embedding-3-small";
  const apiKey = process.env.OPENAI_API_KEY!;

  const res = await fetch(`${baseUrl}/embeddings`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({ model, input: [text] }),
  });

  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`embed API error ${res.status}: ${body.slice(0, 200)}`);
  }

  const json = (await res.json()) as { data: Array<{ embedding: number[] }> };
  const embedding = json.data?.[0]?.embedding;
  if (!Array.isArray(embedding) || embedding.length === 0) {
    throw new Error("embed API returned empty embedding");
  }
  return embedding;
}
